Client/Frontend/Home/BookmarksEditPanel.swift

```diff
@@ -0,0 +1,335 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import UIKit
+import Storage
+import Shared
+import XCGLogger
+
+private let log = Logger.browserLogger
+
+enum DetailPanel: Int {
+    case Title = 0
+    case URL
+    case Folder
+    case _numberOfPanels
+}
+
+struct BookmarksEditPanelUX {
+    private static let LabelLeftRightPadding = 20
+    private static let LabelTopBottomPadding = 10
+    private static let TextFieldTopBottomPadding = 5
+}
+
+class BookmarksEditPanel: SiteTableViewController, HomePanel, UITextFieldDelegate {
+    weak var homePanelDelegate: HomePanelDelegate? = nil
+    var source: BookmarksModel?
+    var parentFolders = [BookmarkFolder]()
+    var bookmarkFolder: BookmarkFolder?
+    var bookmark: BookmarkItem?
+    
+    var titleField: UITextField?
+    var urlField: UITextField?
+    var doneButton: UIButton?
+    
+    lazy var deleteButtonCell = ButtonPanelCell(style: UITableViewCellStyle.Default, reuseIdentifier: "ButtonPanelCell")
+    var deleteButton: UIButton!
+    
+    private let BookmarksEditPanelCellIdentifier = "BookmarksEditPanelCellIdentifier"
+    private let BookmarkFolderEditHeaderViewIdentifier = "BookmarkFolderEditHeaderIdentifier"
+    
+    init() {
+        super.init(nibName: nil, bundle: nil)
+        
+        deleteButton = deleteButtonCell.button
+        deleteButton.setTitle(deleteDeleteButtonLabel, forState: UIControlState.Normal)
+        deleteButton.setTitleColor(UIColor.redColor(), forState: UIControlState.Normal)
+        deleteButton.setTitleColor(UIColor.redColor().colorWithAlphaComponent(0.5), forState: UIControlState.Highlighted)
+        deleteButton.setTitleColor(UIColor.grayColor(), forState: UIControlState.Disabled)
+        deleteButton.addTarget(self, action: #selector(BookmarksEditPanel.deletePressed), forControlEvents: .TouchUpInside)
+        
+        self.tableView.registerClass(BookmarkFolderEditTableViewHeader.self, forHeaderFooterViewReuseIdentifier: BookmarkFolderEditHeaderViewIdentifier)
+        self.tableView.registerClass(BookmarksEditPanelCell.self, forCellReuseIdentifier: BookmarksEditPanelCellIdentifier)
```

Comment:
I'm not sure if it matters and is worth looking into but usually it's better to invoke methods on view objects in `viewDidLoad` vs `init`. I think this causes the tableView to load earlier than the VC would like since UIKit lazy-loads it when accessed.